### PR TITLE
Improve appearance of external links in built-in docs

### LINKED ISSUE #9369: Improve in-engine links that will open in a web browser
### Describe the project you are working on

stuff

### Describe the problem or limitation you are having in your project

image

Sometimes in the integrated docs there are links that when clicked, will open an url in the browser (as opposed to another page of the integrated docs). It's not that clear at a glance.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

- make those links blue with an underline appearing on hover,
- add the "external link" icon at the end of the link text,
- maybe add a tooltip when hovering, stating that it will open in the browser, and maybe the url?

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

image

image

idk if the icon should be colored, it looks correct like that to me.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

can't be worked around

### Is there a reason why this should be core and not an add-on in the asset library?

core already

### PR DESCRIPTION
**Pull request by Mickeon**
Closes https://github.com/godotengine/godot-proposals/issues/9369

This PR makes the external links in the built-in class reference:
- Display with an "external link" icon
- Coloured differently than surrounding text
- Without outline unless hovered on
- With a tooltip containing the link

I'm unconvinced on the actual color, though. See also a [RocketChat thread](https://chat.godotengine.org/channel/documentation/thread/YSvnSkorpFYy7Gu8M). The gist of it is that, I feel like the excessive accent colour brings too much attention to external links, and the icon exists too. Links to the **built-in** class reference are more important in the editor and should be more notable, as well.

| Normal Link Color | Saturated Link Color |
| --- | --- |
| thumb-Clipboard - November 20, 2024 8_08 PM | thumb-Clipboard - November 20, 2024 8_13 PM |
| Clipboard - November 20, 2024 8_08 PM | Clipboard - November 20, 2024 8_14 PM

It's also currently not possible to hide the `hint` tag's underline individually. Hence the dotted underline on these links.

### CODE REVIEW

**Comment by tetrapod00** [comment]
I think the bikeshed should be normal link colored.

File: editor/editor_help.cpp [hunk 1]
```diff
@@ -2522,6 +2523,12 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, const C
 			if (tag != "/img") {
 				p_rt->pop();
 			}
+			if (tag == "/url") {
+				p_rt->pop(); // hint
+				p_rt->pop(); // color
+				p_rt->add_text(" ");
+				p_rt->add_image(p_owner_node->get_editor_theme_icon(SNAME("ExternalLink")), 0, doc_font_size, link_color);
+			}
```
Comment (dalexeev) [inline editor/editor_help.cpp:2531]:
```suggestion
			if (tag == "/img") {
				// Nothing to do.
			} else if (tag == "/url") {
				p_rt->pop(); // meta
				p_rt->pop(); // hint
				p_rt->pop(); // color
				p_rt->add_text(" ");
				p_rt->add_image(p_owner_node->get_editor_theme_icon(SNAME("ExternalLink")), 0, doc_font_size, link_color);
			} else {
				p_rt->pop();
			}
```

File: editor/editor_help.cpp [hunk 2]
```diff
@@ -2787,19 +2796,21 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, const C
 		} else if (tag == "rb") {
 			p_rt->add_text("]");
 			pos = brk_end + 1;
-		} else if (tag == "url") {
-			int end = bbcode.find_char('[', brk_end);
-			if (end == -1) {
-				end = bbcode.length();
+		} else if (tag.begins_with("url")) {
```
Comment (dalexeev) [inline editor/editor_help.cpp:2799]: This changes the logic, as `urlabc` will now work too.

```suggestion
		} else if (tag == "url" || tag.begins_with("url=")) {
```

**Review by dalexeev — APPROVED** [review]
The code looks good to me. Note: I haven't tested this.

File: editor/editor_help.cpp [hunk 3]
```diff
@@ -2519,7 +2520,15 @@ static void _add_text_to_rt(const String &p_bbcode, RichTextLabel *p_rt, const C
...
 			pos = brk_end + 1;
-			if (tag != "/img") {
+			if (tag == "/img") {
+				// Nothing to do.
+			} else if (tag == "/url") {
+				p_rt->pop(); // meta
+				p_rt->pop(); // hint
+				p_rt->pop(); // color
+				p_rt->add_text(" ");
+				p_rt->add_image(p_owner_node->get_editor_theme_icon(SNAME("ExternalLink")), 0, doc_font_size, link_color);
```
[the remark this answers is not in the document]
Comment (Mickeon) [reply editor/editor_help.cpp:2530]: Doing this causes the underline to appear underneath the image when hovered on, which looks weird. But not being able to click the icon would be weird too, I suppose

**Review by akien-mga — APPROVED** [review]

**Comment by adamscott** [comment]
Why the dotted underline for the links? (not related with this PR)

**Comment by Mickeon** [comment]
It's because of the `hint` tag which is required for the URL tooltip (unless https://github.com/godotengine/godot/pull/99481 is implemented)

**Comment by akien-mga** [comment]
Ah right, then let's wait for #99481 to be merged first.

**Comment by Mickeon** [comment]
Rebased to use https://github.com/godotengine/godot/pull/99481 instead.
image

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**